### PR TITLE
Added graceful shutdown code examples

### PR DESCRIPTION
**Pull request by slavafomin**
Added graceful shutdown code examples to the Reliability section of the user guide:

---

screenshot-localhost_8080-2021 11 21-16_43_23

---

screenshot-localhost_8080-2021 11 21-16_43_36

---

screenshot-localhost_8080-2021 11 21-19_01_57

---

### CODE REVIEW

File: site/docs/advanced/reliability.md [hunk 1]
```diff
@@ -11,8 +11,134 @@ All errors that should be expected to happen (failing API calls, failing network
...
+```ts
+import { Bot } from "https://deno.land/x/grammy/mod.ts";
+
+const bot = new Bot("<token>");
+
+// Stopping the bot when Node process
+// is about to be terminated
+['SIGTERM', 'SIGINT'].forEach(signal =>
+  process.once(signal, () => bot.stop())
+);
```
Comment (KnorpelSenf) [inline site/docs/advanced/reliability.md:74]: That's not how you can catch OS signals with Deno. Use this:
```suggestion
// Stopping the bot when Deno process
// is about to be terminated
Deno.addSignalListener('SIGTERM', () => bot.stop());
Deno.addSignalListener('SIGINT', () => bot.stop());
```

File: site/docs/advanced/reliability.md [hunk 2]
```diff
@@ -11,8 +11,134 @@ All errors that should be expected to happen (failing API calls, failing network
...
+
+const bot = new Bot("<token>");
+
+const runner = run(bot);
+
+// Stopping the bot when Node process
+// is about to be terminated
+['SIGTERM', 'SIGINT'].forEach(signal =>
+  process.once(signal, () => runner.isRunning() && runner.stop())
+);
```
Comment (KnorpelSenf) [inline site/docs/advanced/reliability.md:100]:
```suggestion
process.once('SIGTERM', () => bot.stop());
process.once('SIGINT', () => bot.stop());
```

File: site/docs/advanced/reliability.md [hunk 3]
```diff
@@ -11,8 +11,134 @@ All errors that should be expected to happen (failing API calls, failing network
...
```
Comment (KnorpelSenf) [inline site/docs/advanced/reliability.md:119]:
```suggestion
process.once('SIGTERM', () => bot.stop());
process.once('SIGINT', () => bot.stop());
```

File: site/docs/advanced/reliability.md [hunk 4]
```diff
@@ -11,8 +11,134 @@ All errors that should be expected to happen (failing API calls, failing network
...
```
Comment (KnorpelSenf) [inline site/docs/advanced/reliability.md:137]:
```suggestion
// Stopping the bot when Deno process
// is about to be terminated
Deno.addSignalListener('SIGTERM', () => bot.stop());
Deno.addSignalListener('SIGINT', () => bot.stop());
```

File: site/docs/plugins/menu.md [hunk 5]
```diff
@@ -5,7 +5,8 @@ Easily create interactive menus.
 ## Introduction
 
 An inline keyboard is an array of buttons underneath a message.
-grammY has [a built-in plugin](./keyboard.md#inline-keyboards) to create basic inline keybaords.
+grammY has [a built-in plugin](./keyboard.md#inline-keyboards)
+to create basic inline keyboards.
```
Comment (KnorpelSenf) [inline site/docs/plugins/menu.md:9]: Please revert this change. The docs are sticking with a convention of having one sentence per line.
```suggestion
grammY has [a built-in plugin](./keyboard.md#inline-keyboards) to create basic inline keyboards.
```

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]
Thanks for the contributions! A few things need to be fixed:

**Comment by KnorpelSenf** [comment]
What do you think about adding a reference to the runner plugin's page that links to the new section for graceful shutdown?

File: site/docs/advanced/reliability.md [hunk 6]
```diff
@@ -11,8 +11,134 @@ All errors that should be expected to happen (failing API calls, failing network
...
+```ts
+import { Bot } from "grammy";
+
+const bot = new Bot("<token>");
+
+// Stopping the bot when Node process
+// is about to be terminated
+['SIGTERM', 'SIGINT'].forEach(signal =>
+  process.once(signal, () => bot.stop())
+);
```
[the remark this answers is not in the document]
Comment (slavafomin) [reply site/docs/advanced/reliability.md]: Sure, no problem.

File: site/docs/advanced/reliability.md (hunk 1 above)
Comment (slavafomin) [reply site/docs/advanced/reliability.md]: Didn't know that, thanks.

File: site/docs/plugins/menu.md (hunk 5 above)
Comment (slavafomin) [reply site/docs/plugins/menu.md:9]: However, I can't understand the reasoning behind this, considering that it's becoming extremely hard to work with the file, cause it's not fitting on the screen, but it's your monastery :)

**Review by KnorpelSenf — APPROVED** [review]
LGTM. Please run the formatting scripts. Once the translations are done, this can be merged.

**Comment by slavafomin** [comment]
> What do you think about adding a reference to the runner plugin's page that links to the new section for graceful shutdown?

Sure.

File: site/docs/plugins/runner.md [hunk 7]
```diff
@@ -158,6 +158,12 @@ grammY runner will resolve all necessary constraints on the fly and block those
 The implementation of this is very efficient.
 It needs constant memory (unless you specify infinite concurrency), and it needs (amortized) constant processing time per update.
 
+## Graceful shutdown
+
+In order for the bot to complete it's work correctly you
+[should signal](/advanced/reliability.html#using-grammy-runner)
+it to stop when the Node process is about to be destroyed.
```
Comment (KnorpelSenf) [inline site/docs/plugins/runner.md:165]: Please stick with one sentence per line again.

This is important because adding or removing words from sentences causes whole paragraphs to reflow, which makes diffs unreadable and PRs hard to review.

I added this convention to the contribution guidelines now.

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]

File: site/docs/plugins/runner.md (hunk 7 above)
Comment (KnorpelSenf) [reply site/docs/plugins/runner.md:165]:
```suggestion
In order for the bot to complete it's work correctly you [should signal](/advanced/reliability.html#using-grammy-runner) it to stop when the Node process is about to be destroyed.
```

**Comment by slavafomin** [comment]
> LGTM. Please run the formatting scripts. Once the translations are done, this can be merged.

I would be glad to do this, but I don't have Deno to do so, sorry.

**Comment by KnorpelSenf** [comment]
> > LGTM. Please run the formatting scripts. Once the translations are done, this can be merged.
> 
> I would be glad to do this, but I don't have Deno to do so, sorry.

Gimme a sec

**Comment by slavafomin** [comment]
By the way, should the user be responsible for monitoring the process signals? Can't the library do it by itself?

Or we can add a helper functions for this, e.g. `runWithGracefulShutdown(bot)` or `runner.useGracefulShutdown()`.

**Comment by KnorpelSenf** [comment]
It is not included because we don't actually know if the user wants to terminate the instance when the signal is received. It is thinkable that the user wants to keep the bot running for some time, e.g. until some operation has completed. This may be the case very often, but if every used dependency of a user's code base decides for themselves that they want to handle those signals, it can get really, really hard to find out why e.g. postponing a shutdown does not work.

Better not mess around with this. After all, adding two lines of trivial code is simple enough, and now we even have good docs about it.

**Comment by KnorpelSenf** [comment]
I would be open to add this to the grammY runner as a configuration option to the runner config that lets users **opt-in** into registering signals. I do not see any demand for adding this to the core package.

**Review by KnorpelSenf — APPROVED** [review]
Good after translations are done.

**Comment by WingLim** [comment]
I am working on it, please wait.

**Review by WingLim — APPROVED** [review]
Great job!
